iroh-net/src/net/netmon/actor.rs

```diff
@@ -0,0 +1,257 @@
+                biased;
+                _ = debounce_interval.tick() => {
+                    if let Some(time_jumped) = last_event.take() {
+                        if let Err(err) = self.handle_potential_change(time_jumped).await {
+                            warn!("failed to handle network changes: {:?}", err);
+                        };
+                    }
+                }
+                _ = wall_time_interval.tick() => {
+                    trace!("tick: wall_time_interval");
+                    if self.check_wall_time_advance() {
+                        // Trigger potential change
+                        last_event.replace(true);
+                        debounce_interval.reset_immediately();
+                    }
+                }
+                Ok(_event) = self.mon_receiver.recv_async() => {
+                    trace!("network activity detected");
+                    last_event.replace(false);
+                    debounce_interval.reset_immediately();
+                }
+                Some(msg) = self.actor_receiver.recv() => match msg {
+                    ActorMessage::Subscribe(callback, s) => {
+                        let token = self.next_callback_token();
+                        self.callbacks.insert(token, Arc::new(callback));
+                        s.send(token).ok();
+                    }
+                    ActorMessage::Unsubscribe(token, s) => {
+                        self.callbacks.remove(&token);
+                        s.send(()).ok();
+                    }
+                },
+                else => {
+                    break;
+                }
+            }
+        }
+    }
+
+    fn next_callback_token(&mut self) -> CallbackToken {
+        let token = CallbackToken(self.callback_token);
+        self.callback_token += 1;
+        token
+    }
+
+    async fn handle_potential_change(&mut self, time_jumped: bool) -> Result<()> {
+        info!("potential change");
+
+        let new_state = State::new().await;
+        let old_state = &self.interface_state;
+
+        // No major changes, continue on
+        if !time_jumped && old_state == &new_state {
+            debug!("no changes detected");
+            return Ok(());
+        }
+
+        let mut is_major = is_major_change(old_state, &new_state);
+        // Check for time jumps
+        if !is_major && time_jumped {
+            is_major = true;
+        }
+
+        if is_major {
+            self.interface_state = new_state;
+        }
+
+        debug!("triggering {} callbacks", self.callbacks.len());
+        for cb in self.callbacks.values() {
+            let cb = cb.clone();
+            tokio::task::spawn(async move {
+                cb(is_major).await;
+            });
+        }
+
+        Ok(())
+    }
+
+    /// Reports whether wall time jumped more than 150%
+    /// of `POLL_WALL_TIME_INTERVAL`, indicating we probably just came out of sleep.
+    fn check_wall_time_advance(&mut self) -> bool {
```

Comment:
tailscale mentions this being expensive on mobile and not really useful since phones don't "sleep" as laptops. Do you think we should do it anyway?

Reply:
you are right, I forgot to disable this for mobile